fl4health/clients/flash_client.py

```diff
@@ -0,0 +1,100 @@
+from logging import INFO
+from pathlib import Path
+from typing import Dict, Optional, Sequence, Tuple, Union
+
+import torch
+from flwr.common.logger import log
+from flwr.common.typing import Config, Scalar
+
+from fl4health.checkpointing.client_module import ClientCheckpointModule
+from fl4health.clients.basic_client import BasicClient
+from fl4health.utils.losses import LossMeterType
+from fl4health.utils.metrics import Metric
+
+
+class FlashClient(BasicClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[ClientCheckpointModule] = None,
+    ) -> None:
+        """
+        This client is used to perform client-side training associated with the Flash method described in
+        https://proceedings.mlr.press/v202/panchal23a/panchal23a.pdf. Flash is designed to handle statistical
+        heterogeneity and concept drift in federated learning through client-side early stopping and server-side
+        drift-aware adaptive optimization.
+
+        Args:
+            data_path (Path): Path to the data directory.
+            metrics (Sequence[Metric]): List of metrics to be used for evaluation.
+            device (torch.device): Device to be used for training.
+            loss_meter_type (LossMeterType, optional):
+                Type of loss meter to be used. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[ClientCheckpointModule], optional): Checkpointer module defining when and how to do
+                checkpointing during client-side training. No checkpointing is done if not provided. Defaults to None.
+        """
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+        # gamma: Threshold for early stopping based on the change in validation loss.
+        self.gamma: Optional[float] = None
+
+    def process_config(self, config: Config) -> Tuple[Union[int, None], Union[int, None], int, bool]:
+        local_epochs, local_steps, current_server_round, evaluate_after_fit = super().process_config(config)
+        if local_steps is not None:
+            raise ValueError(
```

Comment:
It is too bad that the method is not applicable for train by steps since one of the benchmarks we have been using (FLamby) uses train by steps in the experimental setup. I guess the main issue is that there would be too much variation in the per step losses? Or is there also other factors that do not make it applicable? Just curious for my understanding :)

Reply:
That is quite unfortunate :( one thing is as you mentioned the variation in the per step losses, another thing David mentioned was that FLASH uses validation for these early stopping criteria, which is quite computationally expensive , so it is disabled for now.